### PR TITLE
Use spack.store.use_store consistently in unit tests

### PR DESCRIPTION
**Pull request by alalazo**
#21442 missed a few places in unit tests where the store was monkeypatched within tests. This may cause spurious failures, since monkeypatching was not done consistently over all the global variables in `spack.store`.

Other modifications:

- [x] `spack.store.use_store` has been updated to also swap `spack.store.db`, `spack.store.layout`, `spack.store.root` and `spack.store.unpadded_root`
- [x] `spack.store.store` is serialized in `subprocess_context.py` on Python > 3.8 + MacOS

### CODE REVIEW

**Comment by alalazo** [comment]
@tgamblin This is a follow-up on #21442 In that PR I didn't notice we have tests that monkey patch `spack.store.*`. This is fixed here. Getting tests working correctly on MacOS + Python 3.8+ required serializing also the `store`, so it would be great if @scheibelp could take a look at least at the last commit in the PR.

File: lib/spack/spack/binary_distribution.py
Comment (tgamblin) [inline lib/spack/spack/binary_distribution.py:1130]: There was some kind of circular import issue on certain platforms that led to this being nested here.  Can you look at the blame for these nested `sbang` imports and see if you can figure out whether it's really been resolved?  I think @eugeneswalker and @scheibelp should comment.

File: lib/spack/spack/binary_distribution.py
Comment (tgamblin) [inline lib/spack/spack/binary_distribution.py:1144]: see above comment on circular imports.

**Review by tgamblin — DISMISSED** [review]
I have one request for comment, and I think @scheibelp should review this.

File: lib/spack/spack/binary_distribution.py
Comment (alalazo) [reply lib/spack/spack/binary_distribution.py:1130]: The imports are still nested, but they are done at the top of the function. This is because imports act like a local assignment to the `spack` variable and using any `spack.*` module or function before the nested import will fail with weird errors. Checking `git blame` right now.

Comment (alalazo) [reply lib/spack/spack/binary_distribution.py:1130]: Only reference found for these imports https://github.com/spack/spack/pull/20768#discussion_r555265884

File: lib/spack/spack/store.py
Comment (scheibelp) [inline lib/spack/spack/store.py]: Put these on separate lines

File: lib/spack/spack/test/directory_layout.py
Comment (scheibelp) [inline lib/spack/spack/test/directory_layout.py]: I'm think this should be renamed. The docstring should probably be updated to mention that this is a thin wrapper around `use_store` which just extracts properties for convenience in these unit tests.

File: lib/spack/spack/test/directory_layout.py
Comment (scheibelp) [inline lib/spack/spack/test/directory_layout.py]: I think `yield s.layout, s.root` would be clearer

File: lib/spack/spack/test/conftest.py
Comment (scheibelp) [inline lib/spack/spack/test/conftest.py]: Can this fixture just use the `test_store` fixture without manually constructing a `store` itself.

File: lib/spack/spack/test/conftest.py
Comment (scheibelp) [inline lib/spack/spack/test/conftest.py:566]: Why is the `mock_db.write_transaction` call removed here?

File: lib/spack/spack/test/config_values.py
Comment (scheibelp) [inline lib/spack/spack/test/config_values.py:43]: Can you add a comment that 

```
store = spack.store._store()
with spack.store.use_store(store):
```

is for reinitializing the `directory_layout` with the new prefix settings (i.e. with the shortened hash)?

File: lib/spack/spack/subprocess_context.py
Comment (scheibelp) [inline lib/spack/spack/subprocess_context.py]: This appeared to set `store.db` and `store` without addressing the TODO above, so I'm trying to figure out why tests like `test_installed_upstream` are still passing (they monkeypatch `store.db`, which is undone by this) - my current guess is that the prefixes are determined at the time of the installation and that particular test is only checking prefixes rather than doing DB queries. Overall though those tests are made somewhat fragile by this update and I think this PR will need to address that fragility somehow (e.g. by removing the monkeypatching calls in `test/install.py`).

**Review by scheibelp — CHANGES_REQUESTED** [review]
I have a number of small requests and a larger concern about the consistency of store state when running upstream tests in `test/install.py`:

> This appeared to set store.db and store without addressing the TODO above, so I'm trying to figure out why tests like test_installed_upstream are still passing

I don't have a ready solution for that but we can brainstorm how to deal with it if you'd like.

File: lib/spack/spack/test/conftest.py
Comment (alalazo) [reply lib/spack/spack/test/conftest.py]: Good catch, I'll have a look at factoring that better. There are multiple tests that basically need a temporary store and define their own fixture for that. We have for instance `test_store` in `test/database.py` and `layout_and_dir` above. I think all of them can be unified and the fixture can be moved in `conftest.py`.

File: lib/spack/spack/store.py
Comment (alalazo) [reply lib/spack/spack/store.py]: Done

File: lib/spack/spack/test/config_values.py
Comment (alalazo) [reply lib/spack/spack/test/config_values.py:43]: Comment added, let me know if that's clear enough

File: lib/spack/spack/test/conftest.py
Comment (alalazo) [reply lib/spack/spack/test/conftest.py]: With respect to this and to the other comments on the `layout_and_dir` and `test_store` fixtures: I unified all of them in the `temporary_store` fixture which is now defined in `conftest.py`. Let me know if that addresses all your comments.

File: lib/spack/spack/test/directory_layout.py
Comment (alalazo) [reply lib/spack/spack/test/directory_layout.py]: This has been deleted. The fixture used in its place is `temporary_store`.

File: lib/spack/spack/test/directory_layout.py
Comment (alalazo) [reply lib/spack/spack/test/directory_layout.py]: I'm just yielding the `s` now, and using the corresponding `s.layout` in the tests.

File: lib/spack/spack/test/conftest.py
Comment (alalazo) [reply lib/spack/spack/test/conftest.py:566]: As far as I reconstructed, this context manager was used since #2502.  Initially it was a way to speed-up the fixture by grabbing a write lock early instead of repeatedly take and release the lock for each installation (like Spack would do in its normal operations). This made sense since `database` or `mutable_database` were installing and uninstalling specs at least once per python file. Since some time though the installations are done just once per session, moved to a cache directory and copied wherever they are needed without the additional burden of installing and uninstalling them again. *All this is just to say that we will not lose anything in terms of test performance by removing the context manager.*

Instead, the *necessity* to remove it in this PR comes from the additional serialization I do of attributes related to the store. When we fork a process and we `_serialize` we create an equivalent, but different, store object in the forked process. If a write lock for the same DB is held by the parent process that will prevent the child to proceed.

File: lib/spack/spack/subprocess_context.py
Comment (alalazo) [reply lib/spack/spack/subprocess_context.py]: Calls to `monkey patch.setattr` have been removed in a644293

File: lib/spack/spack/test/conftest.py
Comment (scheibelp) [inline lib/spack/spack/test/conftest.py]: Is it necessary to remove this? I'd assume the `tmpdir` fixture would handle that for you. If it needs to happen, I think it should happen outside of the context manager.

File: lib/spack/spack/test/install.py
Comment (scheibelp) [inline lib/spack/spack/test/install.py:187]: Could you add a docstring to this fixture? Something like:

> Provides a function that installs a specified set of specs to an upstream database. The function returns a store which points to the upstream, as well as the upstream layout (for verifying that dependent installs are using the upstream installs).

**Review by scheibelp — CHANGES_REQUESTED** [review]
I think this refactor is looking excellent. Especially if we can remove all calls which monkeypatch either the store or db, i.e. all calls like:

```
monkeypatch.setattr(spack.store, 'db'
monkeypatch.setattr(spack.store, 'store'
```

I think all such calls have been handled at this point. Calls like 

```
monkeypatch.setattr(spack.store.layout, 'build_packages_path', bpp_path)
```

should be fine.

I have a couple additional requests related to documentation

File: lib/spack/spack/test/install.py
Comment (alalazo) [reply lib/spack/spack/test/install.py:187]: Done in f3ceee6

File: lib/spack/spack/test/conftest.py
Comment (alalazo) [reply lib/spack/spack/test/conftest.py]: There's one test in `bindist.py` that fails without this line, so I moved it outside of the context manager. We can investigate in a later PR why the test is failing. For what is worth I was under the impression that the `tmpdir` had to be cleanup immediately after a succesful test, but it seems a few of them are kept around at any time, see e.g. [here](https://github.com/pytest-dev/pytest/issues/1120)

File: lib/spack/spack/store.py
Comment (scheibelp) [inline lib/spack/spack/store.py:161]: I think you do this to accommodate `subprocess_context.TestState` (these variables are not otherwise referenced) to avoid pickling a `Store` object.

Could you add an explicit `Store.serialize` function and a `Store.deserialize` function like:

```
def serialize(self):
  """
  Return a pickle-able object than can be used to reconstruct a store
  """
  return [self.root, self.unpadded_root, self.projections...]
```

File: lib/spack/spack/subprocess_context.py
Comment (scheibelp) [inline lib/spack/spack/subprocess_context.py]: (question) Out of curiosity, have you tried assigning `self.store = spack.store.store` (instead of saving the properties manually)? I wouldn't be surprised if that failed (i.e. if some object in the hierarchy was not pickle-able) and we already do "manual pickling" for other objects (e.g. `repo_dirs` in this function) so it's not an issue.

**Review by scheibelp — CHANGES_REQUESTED** [review]
I have a final request regarding how the `Store` object is serialized. I think this is good to go after that. Do you agree with the following commit message:

```
Testing: use spack.store.use_store everywhere (#21656)

Keep spack.store.store and spack.store.db consistent in unit tests

* Remove calls to monkeypatch for spack.store.store and spack.store.db:
  tests that used these called one or the other, which lead to
  inconsistencies (the tests passed regardless but were fragile as a
  result)
* Fixtures making use of monkeypatch with mock_store now use the
  updated use_store function, which sets store.store and store.db
  consistently
* subprocess_context.TestState now transfers the serializes and
  restores spack.store.store (without the monkeypatch changes this
  would have created inconsistencies)
```

File: lib/spack/spack/subprocess_context.py
Comment (alalazo) [reply lib/spack/spack/subprocess_context.py]: Yes, I tried that. It failed and, to be honest, I didn't investigate it much since I saw that a `Store` object could be re-created from strings and I thought those were surely pickleable.

**Comment by alalazo** [comment]
The commit message is great, I completely agree with that.

File: lib/spack/spack/store.py
Comment (alalazo) [reply lib/spack/spack/store.py:161]: Done in d3abc7a

**Review by scheibelp — APPROVED** [review]

**Comment by scheibelp** [comment]
Thanks!

**Pull request merged**
